pkg/scheduler/reconcile_on_location.go

```diff
@@ -19,33 +19,239 @@ package scheduler
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
 )
 
 func (c *controller) reconcileOnLocation(ctx context.Context, locKey string) error {
 	logger := klog.FromContext(ctx)
-	ws, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(locKey)
+	lws, _, lName, err := kcpcache.SplitMetaClusterNamespaceKey(locKey)
 	if err != nil {
 		logger.Error(err, "invalid Location key")
 		return err
 	}
-	logger = logger.WithValues("workspace", ws, "location", name)
+	logger = logger.WithValues("locationWorkspace", lws, "location", lName)
 	logger.V(2).Info("reconciling")
 
 	/*
 		On location change:
-		- find all its st(s)
+		1) find all ep(s) that selected loc
+
+		2a) find all its st(s), and update store --- how to get its previous st(s)?
+		2b) find all ep(s) that selecting loc, and update store
 
-		- find all its ep(s), and update store
-		- for each of its obsolete ep, remove all sp(s)
-		- for each of its ongoing ep, update all sp(s)
-		- for each of its new ep, add all sp(s)
+		3a) for each of its obsolete ep, remove all sp(s)
+		3b) for each of its ongoing ep, update all sp(s)
+		3c) for each of its new ep, add all sp(s)
 
 		Need data structure:
 		- map from a location to its eps, say 'epsBySelectedLoc'
+
+		TODO(waltforme): Maybe I can merge 3b) and 3c)
 	*/
 
+	loc, err := c.locationLister.Cluster(lws).Get(lName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		logger.Error(err, "failed to get Location")
+		return err
+	}
+
+	store.l.Lock()
+	defer store.l.Unlock() // TODO(waltforme): Is it safe to shorten the critical section?
+
+	// 1)
+	epsSelectedLoc := store.epsBySelectedLoc[locKey]
+
+	// 2a)
+	stsInLws, err := c.synctargetLister.Cluster(lws).List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "failed to list SyncTargets")
+		return err
+	}
+	stsSelecting, err := filterStsByLoc(stsInLws, loc)
+	if err != nil {
+		logger.Error(err, "failed to find SyncTargets for Location")
+		return err
+	}
+	singles := makeSinglePlacementsForLoc(loc, stsSelecting)
+
+	// TODO: update store
+
+	// 2b)
+	epsAll, err := c.edgePlacementLister.List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "failed to list EdgePlacements in all workspaces")
+		return err
+	}
+	epsFilteredByLoc, err := filterEpsByLoc(epsAll, loc)
+	if err != nil {
+		logger.Error(err, "failed to find EdgePlacements for Location")
+	}
+	epsSelectingLoc := packEpKeys(epsFilteredByLoc)
+
+	store.epsBySelectedLoc[locKey] = epsSelectingLoc
+
+	for ep := range epsSelectedLoc {
+		if _, ok := epsSelectingLoc[ep]; !ok {
+			// 3a)
+			// an (obsolete) ep doesn't select loc anymore
+			// we need to remove all relevant sp(s) from the corresponding sps where 'relevant' means an sp has loc
+			logger.V(1).Info("stop to select Location", "edgePlacement", ep)
+			ws, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(ep)
+			if err != nil {
```

Comment:
This error condition should never arise, because your code controls all the `ep` values, right?  When something impossible happens, I think it is best to not make it abort handling other parts of the problem; better to just log here and move on.

Reply:
Also: if the map keys were `pkg/placement.ExternalName` then there would be no parsing, making the code shorter by both (a) removing the call to the parser and (b) not having to handle the possibility of an error.